### PR TITLE
fix: use okteto api to retrieve dev env endpoints

### PR DESCRIPTION
**Pull request by AdrianPedriza**
## Changes
- use okteto api to retrieve endpoints using `space` query
- use old logic for vanilla or fallbacks if endpoint is not available
- remove external endpoints retrieval when backend is used to retrieve endpoints

## How to test it
- clone okteto/movies
- cd to the repo
- deploy okteto/movies with an external resource (`okteto deploy`)
- check all endpoints are correct comparing them with endpoint in okteto UI when exec `okteto endpoints`
---
### With/without changes
- clone https://github.com/okteto/divert-with-istio-sample
- cd divert-with-istio-sample
- using my own cluster `okteto ctx use <my-context>`. You need to enable istio and an okteto feature flag in order to use virtual services so is easy to use my cluster already prepared.
- run `okteto deploy -f okteto-staging.yml`
- check endpoint in UI (broken because we need an ingress but enough to test the needed logic)
- run `okteto endpoints -f okteto-staging.yml` will show same endpoints than UI using this PR and no endpoints with previous implementation

### CODE REVIEW

**Comment by pchico83** [comment]
@AdrianPedriza we don't need to wait until the backend changes. The code should handle the error and default to the previous behavior. Otherwise, we will be breaking the newer CLIs using old versions of the Okteto chart

**Comment by AdrianPedriza** [comment]
> @AdrianPedriza we don't need to wait until the backend changes. The code should handle the error and default to the previous behavior. Otherwise, we will be breaking the newer CLIs using old versions of the Okteto chart

Absolutely right! I will fallback to the old implementation in that case then

**Comment by AdrianPedriza** [comment]
I'll convert it to draft because I need to call `spaces` endpoint which has the needed info in order to not create a new endpoint in backend

File: pkg/types/interface.go [hunk 1]
```diff
@@ -85,3 +85,8 @@ type KubetokenInterface interface {
 	GetKubeToken(baseURL, namespace string) (KubeTokenResponse, error)
 	CheckService(baseURL, namespace string) error
 }
+
+// EndpointInterface represents the endpoint client
+type EndpointInterface interface {
```
Comment (teresaromero) [inline pkg/types/interface.go:90]:
```suggestion
type EndpointClientInterface interface {
```

File: pkg/endpoints/control.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	"context"
+
+	"github.com/okteto/okteto/pkg/okteto"
+)
+
+type APIControl struct {
+	OktetoClient *okteto.OktetoClient
+}
+
+func NewEndpointControl() (*APIControl, error) {
```
Comment (jLopezbarb) [inline pkg/endpoints/control.go:13]: Should we add the minimum interface as parameter? That way we can test it from the calling function and be able to test it from the calling function.

File: cmd/deploy/endpoints.go [hunk 3]
```diff
@@ -180,23 +195,33 @@ func (eg *EndpointGetter) getEndpoints(ctx context.Context, opts *EndpointsOptio
...
-		return nil, err
-	}
 
-	externalEps, err := eg.endpointControl.List(ctx, opts.Namespace, labelSelector)
-	if err != nil {
-		return nil, err
-	}
+	var eps []string
+	var err error
+	if okteto.Context().IsOkteto {
```
Comment (jLopezbarb) [inline cmd/deploy/endpoints.go:201]: I think this should be an interface. This way we can encapsulate it into a single function making it more readable and the resulting test would be more readable. There is no need if I'm on a Vanilla cluster to have access to `endpointControl` and `externalResourceControl`WDYT?

File: pkg/okteto/endpoint.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	err := query(ctx, &queryStruct, variables, c.client)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoints := make([]string, 0)
+	for _, deployment := range queryStruct.Response.Deployments {
+		if deployment.DeployedBy == graphql.String(deployedBy) {
```
Comment (jLopezbarb) [inline pkg/okteto/endpoint.go:55]: NIT: Given that all of them are components. Should we encapsulate into a function so we don't duplicate code?

File: cmd/deploy/endpoints.go [hunk 5]
```diff
@@ -180,23 +195,33 @@ func (eg *EndpointGetter) getEndpoints(ctx context.Context, opts *EndpointsOptio
...
+		if err != nil {
+			eps, err = eg.getEndpointsInStandaloneMode(ctx, opts, labelSelector, eg.K8sClientProvider)
+			if err != nil {
+				return nil, err
+			}
 
-	for _, externalEp := range externalEps {
-		for _, ep := range externalEp.Endpoints {
-			eps = append(eps, fmt.Sprintf("%s (external)", ep.Url))
+			externalEps, err := eg.externalResourceControl.List(ctx, opts.Namespace, labelSelector)
```
Comment (pchico83) [inline cmd/deploy/endpoints.go:209]: Do we need to query external resources in this case?
If the endpoints query fails, I would show an error and continue. Showing the endpoints is a nice to have, but the command shouldn't fail

**Review by teresaromero — APPROVED** [review]
Functionality tested 👍🏻  deferring approval to address of code feedback

File: cmd/deploy/endpoints.go (hunk 5 above)
Comment (AdrianPedriza) [reply cmd/deploy/endpoints.go]: Yes, that fallback I think is not needed anymore. I did it thinking about creating a new endpoint in the backend and scenarios where okteto chart deployed is not the one containing the new endpoint.. but finally I've reuse `space` query so it shouldn't fail. If the use of the backend endpoint fails I think we should return the error and not fallback. 6ebccf1ff64970e5b09516c5eab7313de63aeb35 WDYT?

Comment (pchico83) [reply cmd/deploy/endpoints.go]: Makes sense

File: pkg/types/interface.go (hunk 1 above)
Comment (AdrianPedriza) [reply pkg/types/interface.go]: 6ebccf1ff64970e5b09516c5eab7313de63aeb35

File: cmd/deploy/endpoints.go (hunk 3 above)
Comment (AdrianPedriza) [reply cmd/deploy/endpoints.go]: e51c9a1cd38266d5e515e36628b2c415714d6e85

File: pkg/okteto/endpoint.go (hunk 4 above)
Comment (AdrianPedriza) [reply pkg/okteto/endpoint.go]: 6e3346e02dfdfc36d5ab324fda5207cdcd1e3f96

**Review by jLopezbarb — APPROVED** [review]

**Review by andreafalzetti — APPROVED** [review]

File: pkg/endpoints/control.go (hunk 2 above)
Comment (AdrianPedriza) [reply pkg/endpoints/control.go]: 52416511b7aff8495f32628b730d7ac78a10671e

**Pull request merged**